Hide terminal cursor

arp242 opened the pull request:
This hides the terminal cursor on startup.

IMHO it doesn't look very nice in the terminal output:

![screenshot_2021-07-11-13-08-06_border](https://user-images.githubusercontent.com/1032692/125183990-fde57f80-e24c-11eb-981d-7e2a1827a1d5.png)

Especially with `--term-osd-bar` it looks pretty meh:

![screenshot_2021-07-11-13-08-43_border](https://user-images.githubusercontent.com/1032692/125183991-ffaf4300-e24c-11eb-88c2-5ae401bca3e9.png)

And even more so if you use something like `--term-osd-bar-chars=┣━╉─┤`:

![screenshot_2021-07-11-13-09-49_border](https://user-images.githubusercontent.com/1032692/125183992-0047d980-e24d-11eb-9477-97dcf08c8408.png)

Not sure if:

1. This is really the best location? 
2. There should be an option for this?
3. How well this works on Windows? I just have a Linux laptop, so I
   can't really test Windows, so I just added a no-op for it.

arp242 commented on osdep/terminal-unix.c:
```diff
@@ -514,6 +518,15 @@ void terminal_get_size2(int *rows, int *cols, int *px_width, int *px_height)
+    if (visible)
+        printf("\e[?25h");
+    else
+        printf("\e[?25l");
```
Note: I read in the contributor document to always add braces, but this file doesn't, and I figured that the style of a particular file takes precedence over the general style.

avih commented:
Personally I'm not fond at all of hiding the cursor by default.

Also, `\033[?25{h,l}` is not vt102, and unsupported for instance on netBSD and Solaris (illumos).

You could instead use a script like so (untested):
```lua
print("\033[25l")
mp.register_event("shutdown", function() print("\033[25h") end)
```

Or just wrap your mpv command in a shell script which disabled-before/enables-after if you really care about the interrupts/bg cases.

arp242 commented:
Yeah, I thought about a wrapper as well; but I figured this might be useful in general.

Different terminals is always an issue; although I don't think it should do any harm? I looked for some terminfo code, but it looks like mpv currently doesn't do anything with that as far as I could find.

It's not a huge thing, just something that's been low-key bothering for a while me and had some time to spare today, and IMH(humble)O it's a small but nice improvement, but also not worth spending ages discussing if it turns out other people don't like it either.

avih commented:
> although I don't think it should do any harm?

Not that I could tell. On netBSD it seem to move the cursor to the beginning of the line, and on illumos it seems to be ignored. That's the TTY console. Obviously if you run mpv from `xterm` or other modern terminals then these escapes are likely to be recognized.

> I looked for some terminfo code, but it looks like mpv currently doesn't do anything with that as far as I could find

IIRC mpv uses only standard terminal codes by default, i.e. which are supported by vt102 or some such and should work on any terminal emulator, though I didn't try to confirm this.

FWIW, I don't think it would be a bad behavior to hide it, but default behavior and terminals are sensitive subjects, so I think personally I'd prefer it as an option, but  I don't know if it would be a wise idea to read option values from a signal handler...

avih commented:
> but I don't know if it would be a wise idea to read option values from a signal handler...

Also, `printf` is not re-entrant and shouldn't be used in a signal handler... https://stackoverflow.com/questions/16891019/how-to-avoid-using-printf-in-a-signal-handler

ericonr commented on osdep/terminal-unix.c:
```diff
@@ -390,6 +393,7 @@ static void close_tty(void)
+    terminal_show_cursor(true);
     do_deactivate_getch2();
 
     (void)write(death_pipe[1], &(char){1}, 1);
```
To avoid calling `printf` in the signal handler, could use whatever event is triggered by writing to the `death_pipe` instead.

ericonr reviewed (commented):
Should wait for someone to complain about it, of course, but if someone likes cursor-less terminals, this would interfere with their setup (though they probably have to carry workarounds for many pieces of software already, so it hardly matters).

ValZapod commented:
It actually looks pretty cool on Windows Terminal UWP app with powershell 7. Because the cursor does not look like inserted one.

arp242 commented:
I changed the `printf()` to `write()` (writing to `death_pipe` doesn't seem to work? 🤔), and I added the `--term-cursor-hide` option for this. This gets passed to the `terminal_setup()` so all the terminal logic can be in `terminal-unix.c`.

avih commented:
> I changed the `printf()` to `write()`

TBH, I really prefer to not touch the signal handler at all. Yes, it won't work on ^Z, but that's a minor issue which those who enable this option should be aware of, if they intend to suspend mpv or put it at the bg.

Also, your code writes the escape sequence even if stdout is not a TTY - which is wrong.

Overall, personally I just don't like this amount of changes only for this option, especially when - after we remove it from the signal handler - it could be implemented with a two-lines mpv lua script or shell wrapper (regardless, it's not a bad idea to add `\e[?25h` to your `PS1`).

So I'm not going to merge this feature. Apologies for making it look as if I will after some changes.

If another mpv dev wants to merge it (after you squash the changes), no one stops them.

arp242 commented:
> Apologies for making it look as if I will after some changes.

No worries; I understand. Thanks for looking at it :-)

At any rate, I mucked around with Lua and this is what I came up with – in case anyone ends up where in an internet search:

```lua
local io    = require "io"
local utils = require "mp.utils"

-- Use io.write as we don't want mpv to show a [martin]: message.
io.write("\x1b[?25l")
mp.register_event("shutdown", function() io.write("\x1b[?25h") end)

-- To allow mapping ^Z:
--    mpv() { stty susp ''; command mpv $@; stty susp '^Z'; }
mp.add_key_binding("ctrl+z", "c-z", function()
	-- Pause before suspending; otherwise mpv will half-read some frame (or some
	-- such) and audio/video is corrupted for a second on resume. 
	mp.command('set pause yes')
	io.write("\x1b[?25h\n")

	mp.add_timeout(1, function()
		io.write("\x1b[?25l\n")
		mp.command('set pause no')
	end)

	utils.subprocess({args = {'kill', '-TSTP', tostring(utils.getpid())}})
end)
```

The suspend/resume stuff is a bit hacky, but it seems to work fine, and also solves another long-time issue: on resume the mpv buffer is frequently corrupted for a few seconds (this was already an issue with mplayer back in the day).

Mapping ^Z won't currently work though, but #9005 should fix that.

avih commented:
> No worries; I understand. Thanks for looking at it :-)

Thanks.

> The suspend/resume stuff is a bit hacky, but it seems to work fine

Cute! :)
